# Project: Classify Medium Articles with Embeddings

Let’s upgrade our **text classification model** as well by leveraging **sentence embeddings**. The scope of the project is to build a text classification model (a simple logistic regression) leveraging sentence embeddings, capable of distinguishing **whether a text is about data science or not**.

## Install and Import Libraries

In [ ]:
%pip install datasets sentence-transformers
%pip install sentence_transformers

In [1]:
from huggingface_hub import hf_hub_download

import pandas as pd
import numpy as np
from sentence_transformers import SentenceTransformer
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import (classification_report, confusion_matrix,ConfusionMatrixDisplay)

d:\anaconda\lib\site-packages\transformers\utils\generic.py:441: UserWarning: torch.utils._pytree._register_pytree_node is deprecated. Please use torch.utils._pytree.register_pytree_node instead.
  _torch_pytree._register_pytree_node(


## Download the Dataset

Download the dataset of [Medium articles from the Hugging Face Hub](https://huggingface.co/datasets/fabiochiu/medium-articles).

In [2]:
df_articles = pd.read_csv(
  hf_hub_download("fabiochiu/medium-articles", repo_type="dataset", filename="medium_articles.csv")
)

df_articles.head()

d:\anaconda\lib\site-packages\huggingface_hub\file_download.py:133: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\zhongmeiqi\.cache\huggingface\hub. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to see activate developer mode, see this article: https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development
  warnings.warn(message)


title  \
0               Mental Note Vol. 24   
1         Your Brain On Coronavirus   
2                    Mind Your Nose   
3          The 4 Purposes of Dreams   
4  Surviving a Rod Through the Head   

                                                text  \
0  Photo by Josh Riemer on Unsplash\n\nMerry Chri...   
1  Your Brain On Coronavirus\n\nA guide to the cu...   
2  Mind Your Nose\n\nHow smell training can chang...   
3  Passionate about the synergy between science a...   
4  You’ve heard of him, haven’t you? Phineas Gage...   

                                                 url                 authors  \
0  https://medium.com/invisible-illness/mental-no...            ['Ryan Fan']   
1  https://medium.com/age-of-awareness/how-the-pa...       ['Simon Spichak']   
2  https://medium.com/neodotlife/mind-your-nose-f...                      []   
3  https://medium.com/science-for-real/the-4-purp...  ['Eshan Samaranayake']   
4  https://medium.com/live-your-life-on-purpose/s...        ['Rishav Sinha']   

                          timestamp  \
0  2020-12-26 03:38:10.479000+00:00   
1  2020-09-23 22:10:17.126000+00:00   
2  2020-10-10 20:17:37.132000+00:00   
3  2020-12-21 16:05:19.524000+00:00   
4  2020-02-26 00:01:01.576000+00:00   

                                                tags  
0  ['Mental Health', 'Health', 'Psychology', 'Sci...  
1  ['Mental Health', 'Coronavirus', 'Science', 'P...  
2  ['Biotechnology', 'Neuroscience', 'Brain', 'We...  
3  ['Health', 'Neuroscience', 'Mental Health', 'P...  
4  ['Brain', 'Health', 'Development', 'Psychology...

## Text Preprocessing and Train/Test Split

First, we concatenate the title and the text content of each article, creating the `full_text` column. We also create the `is_data_science` column, which indicates whether the article has the <font color="blue">“Data Science”</font>tag.

In [3]:
# create two columns:
# - full_text: contains the concatenation of the title and the text of the article.
# - is_data_science: a boolean which is True if the article has the "Data Science" tag
df_articles["is_data_science"] = df_articles["tags"] \
  .apply(lambda tags_list: "Data Science" in tags_list)
df_articles["full_text"] = df_articles["title"] + " " + df_articles["text"]
df_articles.head()

title  \
0               Mental Note Vol. 24   
1         Your Brain On Coronavirus   
2                    Mind Your Nose   
3          The 4 Purposes of Dreams   
4  Surviving a Rod Through the Head   

                                                text  \
0  Photo by Josh Riemer on Unsplash\n\nMerry Chri...   
1  Your Brain On Coronavirus\n\nA guide to the cu...   
2  Mind Your Nose\n\nHow smell training can chang...   
3  Passionate about the synergy between science a...   
4  You’ve heard of him, haven’t you? Phineas Gage...   

                                                 url                 authors  \
0  https://medium.com/invisible-illness/mental-no...            ['Ryan Fan']   
1  https://medium.com/age-of-awareness/how-the-pa...       ['Simon Spichak']   
2  https://medium.com/neodotlife/mind-your-nose-f...                      []   
3  https://medium.com/science-for-real/the-4-purp...  ['Eshan Samaranayake']   
4  https://medium.com/live-your-life-on-purpose/s...        ['Rishav Sinha']   

                          timestamp  \
0  2020-12-26 03:38:10.479000+00:00   
1  2020-09-23 22:10:17.126000+00:00   
2  2020-10-10 20:17:37.132000+00:00   
3  2020-12-21 16:05:19.524000+00:00   
4  2020-02-26 00:01:01.576000+00:00   

                                                tags  is_data_science  \
0  ['Mental Health', 'Health', 'Psychology', 'Sci...            False   
1  ['Mental Health', 'Coronavirus', 'Science', 'P...            False   
2  ['Biotechnology', 'Neuroscience', 'Brain', 'We...            False   
3  ['Health', 'Neuroscience', 'Mental Health', 'P...            False   
4  ['Brain', 'Health', 'Development', 'Psychology...            False   

                                           full_text  
0  Mental Note Vol. 24 Photo by Josh Riemer on Un...  
1  Your Brain On Coronavirus Your Brain On Corona...  
2  Mind Your Nose Mind Your Nose\n\nHow smell tra...  
3  The 4 Purposes of Dreams Passionate about the ...  
4  Surviving a Rod Through the Head You’ve heard ...

Let’s then keep only `1,000` samples of articles with the “Data Science” tag and 1,000 samples without it.

In [4]:
# sample 1000 articles is_data_science = True and 1000 articles with
# is_data_science = False
df = pd.concat([
    df_articles[df_articles["is_data_science"]].sample(n=1000),
    df_articles[~df_articles["is_data_science"]].sample(n=1000)
])

We download a sentence embeddings model called `all-MiniLM-L6-v2`

In [6]:
# download the sentence embeddings model
embedder = SentenceTransformer('all-MiniLM-L6-v2')

… and then use it to generate an embedding for each article in the dataset using the `full_text` column.

In [7]:
# embed article texts
corpus = df["full_text"].values
corpus_embeddings = embedder.encode(corpus)
print(corpus_embeddings.shape)

(2000, 384)


We now have 2,000 embeddings (1,000 for articles with the “Data Science” tag, 1,000 for articles without it), each one with 384 dimensions (which is the number of dimensions of the embeddings produced with the specific `all-MiniLM-L6-v2` model).

Let’s split the articles into training set and test set.

In [8]:
# train/test split
X = corpus_embeddings
y = df["is_data_science"]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2,
                                                    random_state=42, stratify=y)

## Model Training and Evaluation

In this part, Let’s train a `LogisticRegression` model on the training set. And then, we can produce the predictions on the test set and use the `classification_report` utility function from `sklearn.metrics` to quickly see metrics like precision, recall, and F1 score.

This part above requires you to build the code at **Google Colab** yourself. Just try your best! Good luck!